File: generate_qa_pairs.js
const fs = require('fs');
const path = require('path');
const csv = require('csv-parser');

// Configuration
const INPUT_CSV = '500_Synthetic_Interviews_Data.csv'; // Input CSV file
const OUTPUT_DIR = 'output'; // Output directory
const PAIRS_PER_FILE = 5000; // Number of QA pairs per output file

// Ensure output directory exists
function ensureOutputDir() {
  if (!fs.existsSync(OUTPUT_DIR)) {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
  }
}

// Process CSV file and generate QA pairs
async function processCSV(csvFilePath) {
  return new Promise((resolve, reject) => {
    const qaPairs = [];
    
    fs.createReadStream(csvFilePath)
      .pipe(csv())
      .on('data', (row) => {
        // Extract all QA pairs from this row
        for (let i = 1; i <= 33; i++) {
          const questionKey = `Q${i}`;
          const answerKey = `A${i}`;
          
          if (row[questionKey] && row[answerKey]) {
            const qaPair = {
              prompt: row[questionKey],
              completion: row[answerKey]
            };
            qaPairs.push(qaPair);
          }
        }
      })
      .on('end', () => {
        resolve(qaPairs);
      })
      .on('error', (error) => {
        reject(error);
      });
  });
}

// Write QA pairs to output files
function writeOutputFiles(qaPairs) {
  const totalPairs = qaPairs.length;
  const numFiles = Math.ceil(totalPairs / PAIRS_PER_FILE);
  
  console.log(`Writing ${totalPairs} QA pairs to ${numFiles} files...`);
  
  for (let fileIdx = 0; fileIdx < numFiles; fileIdx++) {
    const startIdx = fileIdx * PAIRS_PER_FILE;
    const endIdx = Math.min(startIdx + PAIRS_PER_FILE, totalPairs);
    
    const outputFile = path.join(OUTPUT_DIR, `qa_pairs_part_${fileIdx + 1}.jsonl`);
    
    const fileContent = qaPairs
      .slice(startIdx, endIdx)
      .map(pair => JSON.stringify(pair))
      .join('\n');
    
    fs.writeFileSync(outputFile, fileContent, 'utf8');
    
    console.log(`Created ${outputFile} with ${endIdx - startIdx} QA pairs`);
  }
}

// Main function
async function main() {
  console.log(`Processing CSV file: ${INPUT_CSV}`);
  ensureOutputDir();
  
  try {
    const qaPairs = await processCSV(INPUT_CSV);
    console.log(`Extracted ${qaPairs.length} QA pairs from CSV`);
    
    writeOutputFiles(qaPairs);
    console.log('Done!');
  } catch (error) {
    console.error('Error processing CSV:', error);
  }
}

// Run the script
main();

File: generate_qa_pairs.py
#!/usr/bin/env python3
import csv
import json
import os

# Configuration
INPUT_CSV = "500_Synthetic_Interviews_Data.csv"  # Input CSV file
OUTPUT_DIR = "output"  # Output directory
PAIRS_PER_FILE = 5000  # Number of QA pairs per output file

def ensure_output_dir():
    """Create output directory if it doesn't exist"""
    if not os.path.exists(OUTPUT_DIR):
        os.makedirs(OUTPUT_DIR)

def process_csv(csv_file_path):
    """Process CSV file and generate QA pairs"""
    qa_pairs = []
    
    with open(csv_file_path, 'r', encoding='utf-8') as csvfile:
        reader = csv.DictReader(csvfile)
        for row in reader:
            # Extract all QA pairs from this row
            for i in range(1, 34):  # QA pairs from 1 to 33
                question_key = f"Q{i}"
                answer_key = f"A{i}"
                
                if question_key in row and answer_key in row and row[question_key] and row[answer_key]:
                    qa_pair = {
                        "prompt": row[question_key],
                        "completion": row[answer_key]
                    }
                    qa_pairs.append(qa_pair)
    
    return qa_pairs

def write_output_files(qa_pairs):
    """Write QA pairs to output files"""
    total_pairs = len(qa_pairs)
    num_files = (total_pairs + PAIRS_PER_FILE - 1) // PAIRS_PER_FILE  # Ceiling division
    
    print(f"Writing {total_pairs} QA pairs to {num_files} files...")
    
    for file_idx in range(num_files):
        start_idx = file_idx * PAIRS_PER_FILE
        end_idx = min(start_idx + PAIRS_PER_FILE, total_pairs)
        
        output_file = os.path.join(OUTPUT_DIR, f"qa_pairs_part_{file_idx + 1}.jsonl")
        
        with open(output_file, 'w', encoding='utf-8') as f:
            for idx in range(start_idx, end_idx):
                f.write(json.dumps(qa_pairs[idx], ensure_ascii=False) + '\n')
        
        print(f"Created {output_file} with {end_idx - start_idx} QA pairs")

def main():
    print(f"Processing CSV file: {INPUT_CSV}")
    ensure_output_dir()
    
    qa_pairs = process_csv(INPUT_CSV)
    print(f"Extracted {len(qa_pairs)} QA pairs from CSV")
    
    write_output_files(qa_pairs)
    print("Done!")

if __name__ == "__main__":
    main()
